ddtrace/llmobs/_integrations/langchain.py

```diff
@@ -54,9 +57,63 @@
 SUPPORTED_OPERATIONS = ["llm", "chat", "chain", "embedding", "retrieval", "tool"]
 
 
+def _extract_bound(instance):
+    """
+    Extracts the bound object from a RunnableBinding instance.
+    LangChain will sometimes stick things we trace (chat models) in these bindings,
+    and they will show up as steps in the chain instead of the model instance itself.
+    """
+    if hasattr(instance, "bound"):
+        return instance.bound
+    return instance
+
+
+def _flattened_chain_steps(steps: List[Any], nested: bool = True) -> List[Any]:
+    """
+    Flattens the contents of a chain into non-RunnableBindings and non-RunnableParallel steps.
+    RunnableParrallel steps are extracted and can either be nested into sublists or flattened.
+    """
+    flattened_steps = []
+    for step in steps:
+        step = _extract_bound(step)
+        if hasattr(step, "steps__"):
+            parallel_steps = getattr(step, "steps__", {})
+            flattened_parallel_steps = []
+            for parallel_step in parallel_steps.values():
+                parallel_step = _extract_bound(parallel_step)
+                flattened_parallel_steps.append(parallel_step)
+            if nested:
+                flattened_steps.append(flattened_parallel_steps)
+            else:
+                flattened_steps.extend(flattened_parallel_steps)
+        else:
+            flattened_steps.append(step)
+    return flattened_steps
+
+
+def _is_chain_instance(instance):
+    """Determines if the given instance is a chain instance."""
+    return instance and hasattr(instance, "steps")
+
+
 class LangChainIntegration(BaseLLMIntegration):
     _integration_name = "langchain"
 
+    _spans: Dict[int, Span] = {}
+    """Maps instance ids to spans."""
+
+    _instances: Dict[int, Any] = {}
```

Comment:
let's try to use weakrefs here to avoid the memory leaks

Reply:
reintroduced the weakmap for `_instances` in https://github.com/DataDog/dd-trace-py/pull/12208/commits/377f2bc0b0657b709dc5852ea2938944679b28a4 and manual deleting from `_spans` in https://github.com/DataDog/dd-trace-py/pull/12208/commits/a9f5636d9a9685d454b9eb5e5a7ced10568270ba, which does seem to work!